process/scToProtocol/stakingToPeer.go

```diff
@@ -195,29 +203,58 @@ func (stp *stakingToPeer) updatePeerState(
 	stakingData systemSmartContracts.StakedData,
 	account state.PeerAccountHandler,
 	blsPubKey []byte,
+	nonce uint64,
 ) error {
 
-	err := account.SetRewardAddress(stakingData.RewardAddress)
-	if err != nil {
-		return err
+	var err error
+	if !bytes.Equal(account.GetRewardAddress(), stakingData.RewardAddress) {
+		err = account.SetRewardAddress(stakingData.RewardAddress)
+		if err != nil {
+			return err
+		}
 	}
 
-	err = account.SetBLSPublicKey(blsPubKey)
-	if err != nil {
-		return err
+	if !bytes.Equal(account.GetBLSPublicKey(), blsPubKey) {
+		err = account.SetBLSPublicKey(blsPubKey)
+		if err != nil {
+			return err
+		}
 	}
 
-	err = account.SetStake(stakingData.StakeValue)
-	if err != nil {
-		return err
+	if account.GetStake().Cmp(stakingData.StakeValue) != 0 {
+		err = account.SetStake(stakingData.StakeValue)
+		if err != nil {
+			return err
+		}
 	}
 
-	if stakingData.RegisterNonce != account.GetNonce() {
-		nonceDifference := stakingData.RegisterNonce - account.GetNonce()
-		account.IncreaseNonce(nonceDifference)
-		account.SetNodeInWaitingList(true)
+	isValidator := account.GetList() == string(core.EligibleList) || account.GetList() == string(core.WaitingList)
+	isJailed := stakingData.JailedNonce >= stakingData.UnJailedNonce && stakingData.JailedNonce > 0
+
+	if !isJailed {
+		if stakingData.RegisterNonce == nonce && !isValidator {
+			account.SetListAndIndex(0, string(core.NewList), uint32(stakingData.RegisterNonce))
+			account.SetTempRating(stp.startRating)
+		}
+
+		if stakingData.UnStakedNonce == nonce && account.GetList() != string(core.InactiveList) {
+			account.SetListAndIndex(0, string(core.LeavingList), uint32(stakingData.UnStakedNonce))
```

Comment:
instead of the 0 it could be account.ShardId. The idea with the ListAndIndex if I remember correctly was to have a unique index in the current list: LeavingLIst for example with indexes 1,2,3... Does the unstakeNonce have something to do with the List index?  Multiple validators that unstake in the same block will have the same value here and it should be unique.

Reply:
it is hard to keep a global counter in case of unstakes in the same round, and it depends on luck who gets to be the first among those. Sorting in nodescoordinator will take into consideration validator pub key for deterministic sorting.

Reply:
ok